strings-csharp/SubstringInString/SubstringInString/SubstringSearchMethods.cs

```diff
@@ -0,0 +1,102 @@
+﻿using System.Text.RegularExpressions;
+
+public static partial class SubstringSearchMethods
+{
+    public static List<int> FindAllIndexesWithSubstring(string input, string search)
+    {
+        List<int> indexes = new();
+
+        for (var i = 0; i <= input.Length - search.Length; i++)
+        {
+            if (input.Substring(i, search.Length) == search)
+            {
+                indexes.Add(i);
+            }
+        }
+
+        return indexes;
+    }
+
+    public static void PrintResult(List<int> result)
+    {
+        Console.WriteLine("Indexes where the substring is found: " + string.Join(", ", result));
+    }
+
+    public static List<int> FindAllIndexesWithSpan(string input, string search)
+    {
+        List<int> indexes = new();
+
+        var inputSpan = input.AsSpan();
+        var searchSpan = search.AsSpan();
+
+        for (int i = 0; i <= input.Length - search.Length; i++)
+        {
+            if (inputSpan.Slice(i, search.Length).SequenceEqual(searchSpan))
+            {
+                indexes.Add(i);
+            }
+        }
+
+        return indexes;
+    }
+
+    public static List<int> FindAllIndexesWithIndexOf(string input, string search)
+    {
+        List<int> indexes = new();
+        var startIndex = 0;
+
+        do
+        {
+            startIndex = input.IndexOf(search, startIndex);
+            if (startIndex != -1)
+            {
+                indexes.Add(startIndex);
+                startIndex++;
+            }
+        }
+        while (startIndex != -1);
+
+        return indexes;
+    }
+
+    [GeneratedRegex("ip", RegexOptions.IgnoreCase, "en-US")]
+    private static partial Regex GeneratedRegex();
+    public static List<int> FindAllIndexesWithRegex(string input, string search)
+    {
+        List<int> indexes = new();
+        var regex = GeneratedRegex();
+
+        foreach (Match match in regex.Matches(input))
+        {
+            indexes.Add(match.Index);
+        }
+
+        return indexes;
+    }
+
+    public static List<int> FindAllIndexesWithLINQ(string input, string search)
+    {
+        var indexes = input
+            .Select((c, i) => i <= input.Length - search.Length && input.Substring(i, search.Length) == search ? i : -1)
```

Comment:
We can be using a Span and SequenceEqual here just like we did in the FindIndexesWithSpan method

Reply:
For reference, here is the difference from a benchmark of the two options (the way written here, vs. changing it to use a Span and SequenceEqual):
| Method                     | Mean     | Error     | StdDev    | Ratio | RatioSD | Gen0   | Gen1   | Allocated | Alloc Ratio |
|--------------------------- |---------:|----------:|----------:|------:|--------:|-------:|-------:|----------:|------------:|
| FindAllIndexesWithLINQ     | 2.150 μs | 0.0428 μs | 0.1113 μs |  1.00 |    0.00 | 1.6365 | 0.0019 |    3424 B |        1.00 |
| FindAllIndexesWithLINQSpan | 1.199 μs | 0.0240 μs | 0.0598 μs |  0.56 |    0.05 | 0.1373 |      - |     288 B |        0.08 |

As you can see from the benchmark results, not only is there less overall allocation, the span method is nearly 2x as fast.

For reference, this was my sample data:
```
const string input = "Lorem ipsum dolor sit amet, consectetur adipiscing elip. Duis quis nisip eget sem vehipula accumsan";
const string search = "ip";
```